js/storage.js

```diff
@@ -1283,20 +1283,22 @@ Storage.exportTeam = function (team) {
 			text += '' + curSet.nature + ' Nature' + "  \n";
 		}
 
-		var firstIV = true;
-		for (var j in BattleStatNames) {
-			var val = curSet.ivs[j];
-			if (val === undefined || isNaN(val) || val === 31) continue;
-			if (firstIV) {
-				text += 'IVs: ';
-				firstIV = false;
-			} else {
-				text += ' / ';
+		if (curSet.ivs) {
+			var firstIV = true;
+			for (var j in BattleStatNames) {
+				var val = curSet.ivs[j];
+				if (val === undefined || isNaN(val) || val === 31) continue;
+				if (firstIV) {
+					text += 'IVs: ';
+					firstIV = false;
+				} else {
+					text += ' / ';
+				}
+				text += '' + val + ' ' + BattleStatNames[j];
+			}
+			if (!firstIV) {
+				text += "  \n";
```

Comment:
What kind of value would this give if all IVs were manually set to `31`?

I'm thinking it should say `IVs: 31 all`?

Reply:
Maybe just `IVs: maxed`.

Reply:
Well since we're wiping 31s anyway, why not just treat them as automatic for this particular case instead of introducing a new keyword? That seems to be what [`packTeam` currently does](https://github.com/Zarel/Pokemon-Showdown-Client/blob/8582b6cf52221ccc170956f883ae4293819171c6/js/storage.js#L736-L737)

Reply:
Because we need to distinguish "automatic IVs" from "all 31 IVs".

Reply:
@Zarel I realized that but I didn't think it was worth it to add support for such a rare scenario. ig it's your call though.

Personally, I think `31 all` sounds better than `maxed`, and it makes it clearer what the difference is between that and using the default IVs.

Reply:
I think it's a really common scenario.

I don't think I'm strongly opinionated between `31 all` and `maxed` – but I think they're equally clear about the difference between that and the default IVs.

Reply:
sorry, late to the party here, but not a fan of `31 all`/`maxed` at all. why not just list out the IVs (`31 HP / 31 Atk / ...`)? if we think its that common compared to the default option (which *should* be superior), then we nee to change the default behavior.

adding new keywords here is silly

Reply:
@scheibo Even if automatic IVs is turned off, we will omit IVs with the value 31 while exporting. If all of them are explicitly set to 31 and this code was removed, we would not export any `IVs:` line, so we would read them back in as automatic IVs. We need a keyword to handle this specific case.

Reply:
I'm proposing you just don't omit them. omitted = automatic default, specified = not default Then you don't need to make the parser any more sophisticated.

Reply:
@scheibo If they are not omitted though, then in ie. [this case](https://www.smogon.com/dex/sm/pokemon/scizor/uber/) instead of exporting it as `IVs: 0 Spe`, the user will see `IVs: 31 HP / 31 Atk / 31 Def / 31 SpA / 31 SpD / 0 Spe` all just because they wanted to change a single IV.

Reply:
Can we do:

1) If IVs are not specified at all, do auto behavior
2) If any IVs are specified, the rest that arent specified are 31

This means the set can still be just `IVs: 0 Spe` for `31/31/31/31/31/0`, and you can actually get `31/31/31/31/31/31` behavior with just listing `31 Atk` or whatever (because if any are specified, all get filled in). You can change the implicit fill behavior to instead be the automatic defaults instead of filling with 31s, in which case to specify 'all 31s' you just need to specify 31 on the specific fields that the automatic chooser would get wrong. Either way, you don't need to introduce a new keyword

Reply:
> Can we do:
> 
> 1. If IVs are not specified at all, do auto behavior
> 2. If any IVs are specified, the rest that arent specified are 31
> 
> This means the set can still be just `IVs: 0 Spe` for `31/31/31/31/31/0`, and you can actually get `31/31/31/31/31/31` behavior with just listing `31 Atk` or whatever (because if any are specified, all get filled in).

This is what is implemented in the current version of the PR.

> You can change the implicit fill behavior to instead be the automatic defaults instead of filling with 31s, in which case to specify 'all 31s' you just need to specify 31 on the specific fields that the automatic chooser would get wrong. Either way, you don't need to introduce a new keyword

That was the original approach I took, inferring any IVs that were not explicitly specified (eg. if you put 30 Defense IVs and added Gyro Ball, it would still adjust to 0 Spe, but only say `IVs: 30 Def` in the export). [This](https://github.com/Zarel/Pokemon-Showdown-Client/pull/1333/commits/2415e1deb82418f6903cb07f925f11657f297cbc) is the latest commit before it was rolled back. @Zarel thought it added too much complexity because you have to account for the possibility of `ivs` being only partially filled. I also think it kind of conflicts with expectations a little, if "Use automatic IVs" is unchecked then the implication should be _none_ of them are automatic, as opposed to "they will stay automatic if you don't touch them".

Reply:
First off, I want to apologize for randomly ghosting in here and not following the full commit history/context of this change. I really appreciate your patience and your contribution!

I've slept on this, and I think my main issue is actually with (1). ie/ "if IVs are not specified, do the auto behavior". **I think the default for `IVs` being absent should be 'fill with all 31s' , and nothing sophisticated whatsoever**. That behavior is easy to specify and easy for third party developers to implement. We should not require programs parsing our set export format to ape our 'magic' auto inference logic - if the teambuilder automagically sets a Pokemon with no physical attacks `Atk IV` to 0 or `Happiness` to 0 if they're using `Frustration` or whatever, we **should export that explicitly**. Yes, this means if a Pokemon has `Hidden Power` we should list the `IVs` when they differ from `31` (this means downstream set importers can be super basic!).

For backwards compatibility, this is going to require leaving in additional complexity to deal with legacy behavior, but I think taking on extra complexity here due to a poor past design decision is worth simplifying things for external developers.

I think we're going to want to wait for @Zarel to weigh in here. Sorry for dragging you into this debate :)

Reply:
Sorry, I've had a pretty busy few days.

So one thing we could do is just have an `(auto)` in parentheses next to IVs, if IVs should be auto-inferred. That preserves PS state, while allowing downstream importers to be simpler.

Reply:
I'd prefer that option to `31 all` or `maxed`.

Reply:
@Zarel If I understand you correctly, you're saying have a special keyword for the inferred IVs case rather than the all explicitly 31 case. So if there is no IVs line, then that implicitly means `IVs: 31 HP / 31 Atk / ... / 31 Spe`. But wouldn't that mean there is a change in behavior for ie. a Pokemon with Gyro Ball since we will "see" an explicit 31 Speed, so we can't adjust it to 0 Speed?

Reply:
I would still want a `31 all` or `maxed` keyword. :[ I'm very opposed to actually listing all the 31 IVs.

Reply:
No IVs line would still correspond to auto IVs.

Reply:
In other words, while exporting:

- Export as usual.
- If they're all 31s, `IVs: 31 all`.
- If they're auto and all 31s, leave off the `IVs` line.
- If they're auto and not all 31s, render all the IVs, but add `(auto)` to the end.

While importing:

- If there's no IV line or it ends in `(auto)`, take auto IVs (don't actually read the auto IVs)
- Otherwise, read as normal, using 31 for any skipped IVs

Reply:
@scheibo 

> I think the default for IVs being absent should be 'fill with all 31s' , and nothing sophisticated whatsoever.

Every set exported from the Smogon analyses and all of the pastes on PokePaste omit the `IVs` line under the assumption that the IVs will be automatically adjusted once they are imported into the teambuilder. If we change what it means to not have an `IVs` line, then aren't we changing the meaning of all of those imports and breaking backwards compatibility?

Reply:
> If they're all 31s, `IVs: 31 all.`

I'd have this be no `IVs` line at all

> If they're auto and all 31s, leave off the `IVs` line.

if theyre all 31, how is this not the same as the case above?

> If they're auto and not all 31s, render all the IVs, but add `(auto)` to the end.

I would change this to 'render all the IVs that aren't 31', and don't bother with auto at all.

Basically, remove auto from importing/exporting. Auto as a teambuilder concept, not one which is encoded in the persisted output at all.

Reply:
> I'd have this be no `IVs` line at all

You seem to have missed the comment right above yours for why that's impossible.

> if theyre all 31, how is this not the same as the case above?

With no IV line, edits to the moveset will change the IVs, with `IVs: 31 all`, they wouldn't.

> I would change this to 'render all the IVs that aren't 31', and don't bother with auto at all.
>
> Basically, remove auto from importing/exporting. Auto as a teambuilder concept, not one which is encoded in the persisted output at all.

Auto is a teambuilder concept we expect to be able to preserve when importing and exporting from the teambuilder. People expect to be able to backup/restore their teams without changing them.

Reply:
> You seem to have missed the comment right above yours for why that's impossible. 

It was posted 45s before mine was, so its safe to say I was composing my comment as it was posted and I didn't see it.


> Every set exported from the Smogon analyses and all of the pastes on PokePaste omit the IVs line under the assumption that the IVs will be automatically adjusted once they are imported into the teambuilder.

This is regrettable - a poor decision was made in the past, tying our hands now.